README.md
# client-for-tvdb

[![CI](https://github.com/opacam/client-for-tvdb/workflows/CI/badge.svg?branch=develop)](https://github.com/opacam/client-for-tvdb/actions)
[![codecov](https://codecov.io/gh/opacam/client-for-tvdb/branch/develop/graph/badge.svg)](https://codecov.io/gh/opacam/client-for-tvdb)
[![Python versions](https://img.shields.io/badge/Python-3.6+-brightgreen.svg?style=flat)](https://www.python.org/downloads/)
[![GitHub release](https://img.shields.io/github/release/opacam/client-for-tvdb.svg)](https://gitHub.com/opacam/client-for-tvdb/releases/)
[![GitHub tag](https://img.shields.io/github/tag/opacam/client-for-tvdb.svg)](https://gitHub.com/opacam/client-for-tvdb/tags/)
[![PyPI version fury.io](https://badge.fury.io/py/client-for-tvdb.svg)](https://pypi.python.org/pypi/client-for-tvdb/)
[![GitHub license](https://img.shields.io/github/license/opacam/client-for-tvdb.svg)](https://github.com/opacam/client-for-tvdb/blob/master/LICENSE.md)


A simple client for the [Tvdb API v3](https://api.thetvdb.com/swagger).

## Getting Started

These instructions will get you a copy of the project up and running on your
local machine for development and testing purposes. See deployment for notes on
how to deploy the project on a live system.

### Prerequisites

#### Tvdb Account

You will need an API key from TVDb.com to access the client. To obtain a
key, follow these steps:

- 1. [Register](https://thetvdb.com/auth/register) for and verify an account.
- 2. [Log into](https://thetvdb.com/auth/login) your account.
- 3. [Fill your details](https://thetvdb.com/dashboard/account/apikey/create) to generate a new API key.

#### Python Installation (recommended to use a virtual env)

You also need python >= 3.6 up and running. If your OS does not have the
appropriate python version, you could install [pyenv](https://github.com/pyenv/pyenv) 
and create a virtual environment with the proper python version. Also you will
need an up to date pip installation (version `20.0.2` or greater is our
recommendation). So once you have `pyenv` installed
(see [pyenv install instructions](https://github.com/pyenv/pyenv#installation)), 
make an virtual environment for the project (we will use python version 3.8):

```
pyenv virtualenv 3.8.1 client-for-tvdb
```

Enter inside the python environment we recently created (`client-for-tvdb`):
```
pyenv activate client-for-tvdb
```

Upgrade `pip` package:
```
pip install --upgrade pip
```

Install `poetry` package:
```
pip install poetry
```

### Installing

Once you have the prerequisites installed, you can proceed installing the
project. The project uses an `pyproject.toml` file to manage the installation
(PEP517) and also we will make use of the python package
[poetry](https://github.com/python-poetry/poetry) as our `build-system`
(PEP518). So, to do the install you only need to `cd` to the
project folder:

```
cd client-for-tvdb
```

And run the install of the dependencies via `poetry` command:

```
poetry install
```


## Running API client

To use this tvdb API client, first you must initialize the client with
the proper credentials:

```python
from client_for_tvdb import TvdbClient

tvdb_client = TvdbClient(
    user_name="Your user name",
    user_key="Your user key",
    api_key="Your API key"
)
```

Also you could setup your credentials via environment variables, wrote
in `.env` file which should be located inside the `client_for_tvdb`
module (or you could `export` them):
```
TVDB_USER_NAME=<Your user name>
TVDB_USER_KEY=<Your user key>
TVDB_API_KEY=<Your API key>
```

You can perform the following queries, assuming that you have setup your
credentials via `.env` file:

- To get a list of possible matching tvshows:
  ```python
  from client_for_tvdb import TvdbClient

  tvdb_client = TvdbClient()
  # get a list of dictionaries with tvshows from the TVDB API
  search_result = tvdb_client.search("Game of Thrones")
  ```

- To get only the closest matching tvshow:
  ```python
  from client_for_tvdb import TvdbClient

  tvdb_client = TvdbClient()
  # will return a dictionary
  search_result = tvdb_client.search_closest_matching("Game of Thrones")
  ```

- You also could perform a query supplying a `tvdb_id`
  ```python
  from client_for_tvdb import TvdbClient

  tvdb_client = TvdbClient()
  # will return a dictionary
  search_result = tvdb_client.get_serie_by_id(121361)
  ```

## Running the tests

To run our project tests you can use `pytest` with coverage:

```
PYTHONPATH=. pytest tests/ --cov client_for_tvdb/
```

## Built With

* [Python 3](https://docs.python.org/3/) - The programming language
* [Poetry](https://python-poetry.org/docs/) - Dependency Management

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of
conduct, and the process for submitting pull requests to us.

## Versioning

We use [CalVer](https://calver.org/) for versioning. For the versions available,
see the [tags on this repository](https://github.com/opacam/client-for-tvdb/tags).


## Authors

* **Pol Canelles** - *Initial work* - [opacam](https://github.com/opacam)

See also the list of [contributors](https://github.com/opacam/client-for-tvdb/contributors)
who participated in this project.

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

* [Tvdb API docs](https://api.thetvdb.com/swagger)


CHANGELOG.md
 
# Changelog

## [v2020.05.24](https://github.com/opacam/client-for-tvdb/tree/v2020.05.24) (2020-05-24)
[Full Changelog](https://github.com/opacam/client-for-tvdb/compare/v2020.05.23.post0...v2020.05.24)

**Merged pull requests:**

- :sparkles: Add badge for published PyPI version [\#2](https://github.com/opacam/client-for-tvdb/pull/2) ([opacam](https://github.com/opacam))
- :package: Publish on PyPI when new release created [\#1](https://github.com/opacam/client-for-tvdb/pull/1) ([opacam](https://github.com/opacam))

## [v2020.05.23.post0](https://github.com/opacam/client-for-tvdb/tree/v2020.05.23.post0) (2020-05-24)
[Full Changelog](https://github.com/opacam/client-for-tvdb/compare/v2020.05.23...v2020.05.23.post0)

**General:**

- Add CHANGELOG.md

**Fixed bugs:**

- Update project dependencies to accept python 3.6+
- Make that codecov and gh-actions badges use the develop branch
- Fix coverage report

## [v2020.05.23](https://github.com/opacam/client-for-tvdb/tree/v2020.05.23) (2020-05-24)
[Full Changelog](https://github.com/opacam/client-for-tvdb/compare/22d96468078bbf3cb54bae6e191eead985bb41c1...v2020.05.23)


tests/test_client.py
import json
import os
import pytest
import requests_mock

from pathlib import Path

from client_for_tvdb import client
from client_for_tvdb.config import Config
from client_for_tvdb.end_points import end_points
from client_for_tvdb.exceptions import TvdbClientException

# Disable sqlite cache file
os.environ["TESTING_CLIENT_FOR_TVDB"] = "1"

fake_user_name = "tvdb_user_name"
fake_user_key = "tvdb_user_key"
fake_api_key = "tvdb_api_key"

json_with_token = {"token": "a valid tvdb token"}
json_with_new_token = {"token": "a new valid tvdb token"}

test_files_path = Path(Path(__file__).parent, "files")
json_search_file = Path(test_files_path, "search_result.json")
with open(json_search_file) as output_file:
    json_search_result = json.load(output_file)

simulated_tvdb_result = {
    'aliases': [
        'Game of Thrones - Das Lied von Eis und Feuer',
        'Le Trône de fer',
        '权力的游戏',
        '冰与火之歌 第四季',
        '權力遊戲'
    ],
    'banner': '/banners/graphical/121361-g19.jpg',
    'firstAired': '2010-12-5',
    'id': 121361,
    'image': '/banners/posters/121361-4.jpg',
    'network': 'HBO',
    'overview':
        'Seven noble families fight for control of the mythical land of '
        'Westeros. Friction between the houses leads to full-scale war. '
        'All while a very ancient evil awakens in the farthest north. '
        'Amidst the war, a neglected military order of misfits, the '
        "Night's Watch, is all that stands between the realms of men and "
        'the icy horrors beyond.',
    'poster': '/banners/posters/121361-4.jpg',
    'seriesName': 'Game of Thrones',
    'slug': 'game-of-thrones',
    'status': 'Ended'
}


@pytest.fixture
def mock_config_credentials(monkeypatch):
    monkeypatch.setattr(Config, "user_name", fake_user_name)
    monkeypatch.setattr(Config, "user_key", fake_user_key)
    monkeypatch.setattr(Config, "api_key", fake_api_key)


@pytest.fixture
def mock_config_missing_api(monkeypatch):
    monkeypatch.setattr(Config, "api_key", "")


@pytest.fixture
def mock_tvdb_login(mock_config_credentials):
    with requests_mock.Mocker() as mocked:
        mocked.post(end_points.login, json=json_with_token, status_code=200)
        yield mocked


def test_tvdb_client_login_success(mock_config_credentials):
    # test a straight forward login
    with requests_mock.Mocker() as m:
        m.post(end_points.login, json=json_with_token, status_code=200)
        tvdb_client = client.TvdbClient()
    assert isinstance(tvdb_client, client.TvdbClient) is True
    assert tvdb_client.token == json_with_token["token"]


def test_tvdb_client_login_after_token_refresh(mock_config_credentials):
    # test a login which requires refresh_token
    with requests_mock.Mocker() as m:
        m.post(end_points.login, json=json_with_token, status_code=401)
        m.get(
            end_points.refresh_token,
            json=json_with_new_token,
            status_code=200,
        )
        tvdb_client = client.TvdbClient()
    assert tvdb_client.token == json_with_new_token["token"]


def test_tvdb_client_login_missing_api(mock_config_missing_api):
    with pytest.raises(TvdbClientException) as tvdb_exc:
        client.TvdbClient()
    assert client.CONFIGURE_YOUR_CREDENTIALS_ERROR_MESSAGE in str(
        tvdb_exc.value
    )
    assert client.CONFIGURE_CREDENTIALS_INSTRUCTIONS_MESSAGE in str(
        tvdb_exc.value
    )


def test_tvdb_client_refresh_token(mock_tvdb_login):
    tvdb_client = client.TvdbClient()
    with requests_mock.Mocker() as m:
        m.get(
            end_points.refresh_token,
            json=json_with_new_token,
            status_code=200,
        )
        tvdb_client.refresh_token()
    assert tvdb_client.token == json_with_new_token["token"]

    # test refresh_token error (which should return an status code)
    with requests_mock.Mocker() as m:
        m.get(end_points.refresh_token, status_code=401)
        error_code = tvdb_client.refresh_token()
    assert error_code == 401


def test_tvdb_client_search(mock_tvdb_login):
    tvdb_client = client.TvdbClient()
    with requests_mock.Mocker() as m:
        m.get(
            f"{end_points.search}?name=Game+of+Thrones",
            json=json_search_result,
            status_code=200,
        )
        listed_results = tvdb_client.search("Game of Thrones")
    assert isinstance(listed_results, list) is True
    assert len(listed_results) == 5
    for item in listed_results:
        assert isinstance(item, dict) is True


def test_tvdb_client_search_error(mock_tvdb_login):
    tvdb_client = client.TvdbClient()
    with requests_mock.Mocker() as m:
        m.get(
            f"{end_points.search}?name=Game+of+Thrones",
            status_code=404,
        )
        listed_results = tvdb_client.search("Game of Thrones")
    assert listed_results is None


def test_tvdb_client_search_closest_matching(mock_tvdb_login):
    tvdb_client = client.TvdbClient()
    with requests_mock.Mocker() as mock_search:
        mock_search.get(
            f"{end_points.search}?name=Game+of+Thrones",
            json=json_search_result,
            status_code=200,
        )
        data_dict = tvdb_client.search_closest_matching("Game of Thrones")
    for k, v in data_dict.items():
        assert v == simulated_tvdb_result[k]


def test_tvdb_client_get_serie_by_id(mock_tvdb_login):
    tvdb_client = client.TvdbClient()
    serie_id = 121361

    with requests_mock.Mocker() as mock_get_id:
        mock_get_id.get(
            f"{end_points.series}/{serie_id}",
            json={"data": simulated_tvdb_result},
            status_code=200,
        )
        serie_data = tvdb_client.get_serie_by_id(serie_id)
    for k, v in serie_data.items():
        assert v == simulated_tvdb_result[k]


tests/test_client_for_tvdb.py
from client_for_tvdb import __version__


def test_version_for_module():
    assert __version__ == '2020.05.24'



client_for_tvdb/end_points.py
base_uri = "https://api.thetvdb.com"


class EndPoints:
    """
    A class which define tvdb's api endpoints.
    """

    # Authentication (https://api.thetvdb.com/swagger#/Authentication)
    login = f"{base_uri}/login"
    refresh_token = f"{base_uri}/refresh_token"

    # Search (https://api.thetvdb.com/swagger#/Search)
    search = f"{base_uri}/search/series"

    # Series (https://api.thetvdb.com/swagger#/Series)
    series = f"{base_uri}/series"

    # Episodes (https://api.thetvdb.com/swagger#/Episodes)
    episodes = f"{base_uri}/episodes"

    # Movies (https://api.thetvdb.com/swagger#/Movies)
    movies = f"{base_uri}/movies"

    # Updates (https://api.thetvdb.com/swagger#/Updates)
    updated_query = f"{base_uri}/updated/query"
    updated_query_params = f"{base_uri}/updated/query/params"

    # Languages (https://api.thetvdb.com/swagger#/Languages)
    languages = f"{base_uri}/languages"

    # Users (https://api.thetvdb.com/swagger#/Users)
    user = f"{base_uri}/user"
    user_favorites = f"{base_uri}/user/favorites"
    user_ratings = f"{base_uri}/user/ratings"
    user_ratings_query = f"{base_uri}/user/ratings/query"
    user_ratings_query_params = f"{base_uri}/user/ratings/query/params"


end_points = EndPoints()


client_for_tvdb/client.py
import json
import logging
import os
import requests
import requests_cache

from difflib import get_close_matches
from typing import Union

from client_for_tvdb.config import Config
from client_for_tvdb.end_points import end_points
from client_for_tvdb.exceptions import TvdbClientException

logging.basicConfig(level=logging.DEBUG)
log = logging.getLogger("tvdb")

requests_cache_name = "tvdb_api_cache"

CONFIGURE_YOUR_CREDENTIALS_ERROR_MESSAGE = (
    "TvdbClient couldn't be started due to wrong credentials. Please, review "
    "your settings and configure it properly."
)

CONFIGURE_CREDENTIALS_INSTRUCTIONS_MESSAGE = (
    "You will need an API key from `TVDb.com` to access the client. Please, "
    "Follow the instructions from: "
    "https://github.com/opacam/client-for-tvdb#tvdb-account"
)


class TvdbClient:
    token = None

    def __init__(
            self, user_name: str = "", user_key: str = "", api_key: str = "",
    ) -> None:
        """
        Initialize the tvdb client by setting user credentials, requests
        cached session and tries to login the user account with the supplied
        credentials.
        """
        self.user_name = user_name or Config.user_name
        self.user_key = user_key or Config.user_key
        self.api_key = api_key or Config.api_key
        print(f"self.username is: {self.user_name}")

        self.session = requests_cache.CachedSession(
            # cache will expire after 6 hours
            expire_after=21600,
            backend="sqlite"
            if "TESTING_CLIENT_FOR_TVDB" not in os.environ
            else "memory",
            cache_name=requests_cache_name,
            include_get_headers=True,
        )
        self.session.remove_expired_responses()

        if self.user_name and self.user_key and self.api_key:
            self.login()
        else:
            raise TvdbClientException(
                CONFIGURE_YOUR_CREDENTIALS_ERROR_MESSAGE,
                instructions=CONFIGURE_CREDENTIALS_INSTRUCTIONS_MESSAGE,
            )

    @property
    def session_headers(self) -> dict:
        """Returns a generic headers for perform our tvdb api queries."""
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            # Todo: make language dynamic
            "Accept-Language": "en",
        }
        if self.token:
            headers.update({"Authorization": f"Bearer {self.token}"})
        return headers

    @property
    def user_params(self) -> dict:
        """Returns a dict with the user credentials."""
        params = {"apikey": self.api_key}
        if self.user_key and self.user_key:
            params.update({
                "userkey": self.user_key, "username": self.user_name,
            })
        return params

    def _post(self, endpoint: str) -> requests.Response:
        """Generic `requests.post` calls."""
        response = self.session.post(
            endpoint, headers=self.session_headers, json=self.user_params,
        )
        return response

    def _get(self, endpoint: str, params: dict = None) -> requests.Response:
        """Generic `requests.get` calls."""
        response = self.session.get(
            endpoint, headers=self.session_headers, params=params,
        )
        return response

    def login(self) -> Union[str, None]:
        """
        Login into tvdb account but in case that the login fails we will try
        to refresh the token before giving up.
        """
        log.info(f"Login to tvdb api with user: {self.user_name}")
        response = self._post(end_points.login)
        if response.status_code == 200:
            self.token = json.loads(response.content)["token"]
        elif response.status_code == 401:
            log.error("Cannot login into  account, trying to refresh token...")
            response_code = self.refresh_token()
            if response_code == 200:
                log.info("Ok refreshed token.")
            elif response_code == 401:
                log.error("Seems that your JWT token is missing or expired.")
            else:
                log.error("Invalid credentials and/or API token.")
        else:
            log.error(f"Error when trying to login: {response.content}")
        return self.token

    def refresh_token(self) -> int:
        """
        Refresh the current token set in the module. Returns the the status
        code of the response.
        """
        response = self._get(end_points.refresh_token, params=self.user_params)

        if response.status_code == 200:
            self.token = json.loads(response.content)["token"]
        else:
            log.error(f"Error while refreshing token: {response.content}")
        return response.status_code

    def search(self, name: str) -> Union[list, None]:
        """Given a series name, return a list of possible series."""
        response = self._get(end_points.search, params={"name": name})
        if response.status_code == 200:
            return json.loads(response.content)["data"]
        log.error(f"Error while searching series by name: {response.content}")
        return None

    def search_closest_matching(self, name: str) -> Union[dict, None]:
        """
        Given a series name, search all possible series and return the closest
        matching serie in a dict format.
        """
        series = self.search(name)
        if series is None:
            return None

        # Tries to get the closest matching series
        names = [i["seriesName"] for i in series if i["seriesName"]]
        res = get_close_matches(name, names, n=len(series))
        for serie in series:
            if res and serie["seriesName"] == res[0]:
                log.info(
                    " closest result for tvdb search is: {}".format(
                        serie["seriesName"]
                    )
                )
                return serie

        # we didn't find the closest match, so we take the first result
        return series[0]

    def get_serie_by_id(self, serie_id: int) -> Union[dict, None]:
        """Given a series name, return a list of possible series."""
        response = self._get(f"{end_points.series}/{serie_id}")
        if response.status_code == 200:
            return json.loads(response.content)["data"]
        log.error(f"Error while searching series by name: {response.content}")
        return None
